kernel/kernel-api/src/main/java/io/delta/kernel/internal/TransactionImpl.java

```diff
@@ -151,9 +154,12 @@ private TransactionCommitResult doCommit(
             throws FileAlreadyExistsException {
         List<Row> metadataActions = new ArrayList<>();
         metadataActions.add(createCommitInfoSingleAction(generateCommitAction()));
+        if (shouldUpdateMetadata || isNewTable) {
+            metadataActions.add(createMetadataSingleAction(metadata.toRow()));
+            shouldUpdateMetadata = false;
```

Comment:
why does this need to be reset?

Reply:
Because we provide a way to set the table properties which means we need to reset the metadata even when it's not a new table.

Reply:
I believe the rationale behind adding `shouldUpdateMetadata` is clear but what is unclear is why it needs to reset to `false`? We don't do that for `isNewTable`

Reply:
Because we may try `doCommit` for multiple times. We only need to update metadata when it's updated during resolving the conflicts.